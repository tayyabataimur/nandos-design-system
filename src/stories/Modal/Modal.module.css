@import url("https://fonts.googleapis.com/css2?family=Barlow:ital,wght@0,100;0,200;0,300;0,400;0,500;0,600;0,700;0,800;0,900;1,100;1,200;1,300;1,400;1,500;1,600;1,700;1,800;1,900&display=swap");

.modalOverlay {
  background: #00000099;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 100%;
  position: fixed;
  top: 0;
  width: 100%;
  padding: 1rem;
  z-index: 80;
}

.modal {
  background: #fffef6;
  border-radius: 0.5rem;
  color: #000000;
  display: flex;
  flex-direction: column;
  gap: 2rem;
  padding: 1rem;
  z-index: 10;
}

.modalHeader {
  align-items: flex-start;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 100%;
}

.modalHeading {
  font-family: "Barlow", sans-serif;
  font-weight: 400;
  font-size: 30px;
  line-height: 40.5px;
  margin: 0;
}

.modalBody {
  font-family: "Barlow", sans-serif;
  font-weight: 400;
  font-size: 22px;
  line-height: 34px;
}

@media (min-width: 768px) {
  .modal {
    max-width: 560px;
  }
}
