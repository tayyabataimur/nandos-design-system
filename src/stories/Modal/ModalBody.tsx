import React from "react";
import styles from "./Modal.module.css";
import cx from "classnames";

type ModalBodyProps = {
  children: React.ReactNode;
  additionalClassnames?: string;
};

const ModalBody = ({ children, additionalClassnames }: ModalBodyProps) => {
  return (
    <div className={cx(styles.modalBody, additionalClassnames)}>{children}</div>
  );
};

export default ModalBody;
