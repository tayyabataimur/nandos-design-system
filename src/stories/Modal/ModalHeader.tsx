import React from "react";

import styles from "./Modal.module.css";
import IconButton from "../IconButton";

type ModalHeaderProps = {
  children: React.ReactNode;
  onClose: () => void;
  disableOnClose?: boolean;
};

const ModalHeader = ({
  children,
  onClose,
  disableOnClose = false,
}: ModalHeaderProps) => {
  return (
    <div className={styles.modalHeader}>
      <h2 className={styles.modalHeading}>{children}</h2>
      <IconButton icon="close" onClick={onClose} isDisabled={disableOnClose} />
    </div>
  );
};

export default ModalHeader;
