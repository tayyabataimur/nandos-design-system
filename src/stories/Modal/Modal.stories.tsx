import type { Meta, StoryObj } from "@storybook/react";

import Modal from ".";
import { useState } from "react";
import Button from "../Button";

const meta: Meta<typeof Modal> = {
  component: Modal,
  tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof Modal>;

export const Primary: Story = {
  args: {
    heading: "This is a modal",
    children:
      "This placeholder component can be replaced with your own local component.",
  },
  render: (args) => {
    const [isOpen, setIsOpen] = useState(false);

    const handleOpen = () => setIsOpen(true);
    const handleClose = () => setIsOpen(false);

    return (
      <div>
        <Button onClick={handleOpen}>Open Modal</Button>
        <Modal
          heading={args.heading}
          isOpen={isOpen}
          onClose={handleClose}
          ariaLabelledby="modal"
        >
          {args.children}
        </Modal>
      </div>
    );
  },
};
