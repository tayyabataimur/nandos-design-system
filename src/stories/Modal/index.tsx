import React from "react";
import styles from "./Modal.module.css";
import ModalHeader from "./ModalHeader";
import ModalBody from "./ModalBody";

type ModalProps = {
  /**
   * The heading of your modal
   */
  heading: string;
  /**
   * To show the modal
   */
  isOpen: boolean;
  /**
   * What happens when your modal is closed
   */
  onClose: () => void;
  /**
   * Accessibility label
   */
  ariaLabelledby: string;
  /**
   * What you want to render inside your modal
   */
  children: React.ReactNode;
  /**
   * Disable closing the modal
   */
  disableOnClose?: boolean;
  /**
   * A unique ID for your modal
   */
  id?: string;
};

/**
 *
 * #Overview
 * A modal is a popup window that appears on top of a webpage or application to
 * display information or prompt user interaction. It typically requires user action
 * before returning to the main interface.
 * */

const Modal = ({
  heading,
  isOpen = true,
  onClose,
  ariaLabelledby,
  children,
  disableOnClose = false,
  id,
}: ModalProps) => {
  return (
    (isOpen && (
      <div className={styles.modalOverlay}>
        <div
          className={styles.modal}
          aria-labelledby={ariaLabelledby}
          aria-modal={true}
          role="dialog"
          id={id}
        >
          <ModalHeader onClose={onClose} disableOnClose={disableOnClose}>
            {heading}
          </ModalHeader>
          <ModalBody>{children}</ModalBody>
        </div>
      </div>
    )) ||
    null
  );
};

export default Modal;
