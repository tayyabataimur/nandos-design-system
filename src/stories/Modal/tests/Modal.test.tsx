import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import Modal from '../';

describe('Modal component', () => {
  it('renders correctly when isOpen is true', () => {
    render(
      <Modal heading="Test Modal" isOpen={true} onClose={() => {}} ariaLabelledby="test-heading">
        <div>Modal content</div>
      </Modal>
    );
    expect(screen.getByRole('dialog')).toMatchSnapshot();
    expect(screen.getByText('Test Modal')).toBeInTheDocument();
    expect(screen.getByText('Modal content')).toBeInTheDocument();
  });

  it('does not render when isOpen is false', () => {
    render(
      <Modal heading="Test Modal" isOpen={false} onClose={() => {}} ariaLabelledby="test-heading">
        <div>Modal content</div>
      </Modal>
    );
    expect(screen.queryByText('Test Modal')).not.toBeInTheDocument();
    expect(screen.queryByText('Modal content')).not.toBeInTheDocument();
  });

  it('calls onClose when close button is clicked', () => {
    const onCloseMock = jest.fn();
    render(
      <Modal heading="Test Modal" isOpen={true} onClose={onCloseMock} ariaLabelledby="test-heading">
        <div>Modal content</div>
      </Modal>
    );
    fireEvent.click(screen.getByRole('button'));
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });


  it('renders with disableOnClose set to true', () => {
    render(
      <Modal heading="Test Modal" isOpen={true} onClose={() => {}} ariaLabelledby="test-heading" disableOnClose={true}>
        <div>Modal content</div>
      </Modal>
    );
    expect(screen.getByRole('button')).toBeDisabled();
  });

  it('renders without disableOnClose (defaults to false)', () => {
    render(
      <Modal heading="Test Modal" isOpen={true} onClose={() => {}} ariaLabelledby="test-heading">
        <div>Modal content</div>
      </Modal>
    );
    expect(screen.getByRole('button')).not.toBeDisabled();
  });
});
