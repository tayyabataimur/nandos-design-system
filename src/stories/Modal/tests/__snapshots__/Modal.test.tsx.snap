// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component renders correctly when isOpen is true 1`] = `
<div
  aria-labelledby="test-heading"
  aria-modal="true"
  class="modal"
  role="dialog"
>
  <div
    class="modalHeader"
  >
    <h2
      class="modalHeading"
    >
      Test Modal
    </h2>
    <button
      class="iconButton"
    >
      <svg
        fill="none"
        height="12"
        viewBox="0 0 13 12"
        width="13"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.4 1.20857L11.1915 0L6.40002 4.79143L1.6086 0L0.400024 1.20857L5.19145 6L0.400024 10.7914L1.6086 12L6.40002 7.20857L11.1915 12L12.4 10.7914L7.6086 6L12.4 1.20857Z"
          fill="black"
        />
      </svg>
    </button>
  </div>
  <div
    class="modalBody"
  >
    <div>
      Modal content
    </div>
  </div>
</div>
`;
